### **EDA-3. Проектирование признаков**

#### **I. Введение**

В этом модуле мы поговорим об ещё одном направлении разведывательного анализа данных - **проектировании признаков (Feature Engineering).** Этот процесс также называют **конструированием признаков или разработкой признаков.** Все эти термины описывают процесс создания признаков, которые нужны для обучения модели.

*Но мы же уже собрали данные на этапе сбора данных и подготовили данные на этапе очистки. Для чего повторно создавать части в EDA?*

Как мы выяснили в предыдущих модулях, бывает недостаточно данных, собранных после этапа формирования проблемы. Данные могут быть представлены в «неподобающем» для обучения виде, например в формате даты или времени, большого текста. 

**Пример:** При прогнозировании дефолта клиента банка для создания прогнозной модели очень важно иметь информацию о дате выдачи кредита, плановой и действительной датах погашения. Эти даты дают нам представление о характере погашения кредита: был он выплачен вовремя или раньше срока, с просрочкой. 

В отличие от человека, модель машинного обучения не может проанализировать эти даты. А вот информация о просрочке (разница между плановой и фактической датой погашения) представлена в числовом формате и может сообщить некую информацию модели. Число, равное или меньше **0,** говорит о клиенте как о благонадёжном заёмщике; число больше **0 -** характеризует заёмщика как должника. Такое конструирование признаков помогает дата-сайентисту не утратить часть данных, а также представить их в более «выгодном» свете.

В знакомой нам задаче винных обзоров информация о вкусовых качествах вина также может оказаться важной при его оценке, но не может быть передана в модель машинного обучения, так как этот признак - сплошной текст. Однако мы можем вычленить из текста информацию о положительной или отрицательной оценке вина, превратив её в признаки. Такой признак, скорее всего, оказался бы важным в определении качественных вин. Определить его значимость мы сможем только на этапе отбора признаков, который мы также будем изучать в модулях *EDA*.

Избавляться от перечисленных выше данных было бы неправильно, ведь мы можем лишиться части потенциально важной информации. Проектирование признаков как раз занимается преобразованием такой информации в признаки для будущей модели машинного обучения. 

На этапе проектирования признаков специалисты по данным преследуют такие же цели, как на этапе сбора данных - сконструировать как можно больше признаков, чтобы далее отобрать из них лучшие.

Прежде чем мы приступим к изучению нового материала, необходимо ответить на несколько вопросов.


**1.** В разведывательном анализе данных Feature Engineering - это …

* *конструирование признаков*
* *проектирование признаков*
* *создание признаков*

**2.** Проектирование признаков необходимо для …

* *очистки признаков*

#### **II. Создание признаков**

**Проектирование признаков -** самый творческий процесс во всём цикле жизни создания модели машинного обучения. 

*Он включает в себя несколько этапов:*

* *создание признаков*;
* *преобразование признаков*;
* *отбор признаков*.

В данном модуле мы подробно познакомимся с каждым из этапов проектирования признаков. Этот юнит будет посвящён процессу создания признаков.

Мы уже знакомы с методами создания столбцов в *Pandas* из модуля *PYTHON-11. Базовые приемы работы с данными в Pandas. Юнит 2. Базовые операции со столбцами DataFrame.*

Создание признаков, **feature extraction,** часто именуемое **генерацией признаков,** - творческий процесс извлечения новых признаков путём создания признаков из существующего набора данных или путём использования дополнительной информации. 

Проектирование признаков необходимо для улучшения качества будущей модели: в процессе создания признаков появляются новые, возможно, более качественные, чем исходные.

*Новые признаки могут быть сконструированы двумя способами:*

* с помощью внешних источников данных;
* из существующего набора данных.

**Пример 1:** В колл-центрах часто используются системы автоматического подбора номеров для дозвона (рекомендация). Если клиент во время прошлого созвона сбросил трубку, то автоматическая система должна снижать рекомендацию этого номера. Клиент вряд ли ответит после того, как бросил трубку. 

В выделении этого факта нам поможет создание нового признака «сброс трубки» из информации в существующем датасете. Эта информация может быть получена из уже существующих признаков, например если время звонка составило меньше 10 секунд.

**Пример 2:** Для определения дефолтности клиента банки делают запрос в БКИ (бюро кредитных историй, которое занимается организацией, хранением и предоставлением кредитных отчётов, в которых содержатся сведения о принятых на себя обязательствах по договорам займа и кредита) с целью получения дополнительной информации по клиенту. 

Такая информация о невыплаченных долгах другим организациям может быть решающим фактором в вынесении решения по кредиту. Эту информацию можно получить с использованием внешних источников данных.

Работе с внешними источниками данных будет посвящен следующий юнит.

**Новые признаки можно сконструировать из уже существующего набора данных несколькими способами:**

* разбор категорий; 
* разбор даты и времени;
* разбор числовых признаков;
* разбор текста.

Рассмотрим эти способы на примере задачи винных обзоров. Вспомним задачу винного сомелье - научиться определять рейтинг (качество) вина по его различным характеристикам. В этом модуле для практических заданий мы будем пользоваться подготовленным и очищенным датасетом.

In [2]:
# Импортируем библиотеку Pandas
import pandas as pd

# Используем функцию filterwarnings из библиотеки warnings для отключения предупреждений
import warnings
warnings.filterwarnings('ignore')

Прочитаем подготовленный и очищенный файл с винными обзорами и выведем первые пять строк таблицы с помощью метода **head(),** чтобы убедиться в том, что всё подгрузилось верно:

In [3]:
df_wine_cleared = pd.read_csv('data/wine_cleared.csv')
df_wine_cleared.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points      price           province  \
0                        Vulkà Bianco      87  35.363389  Sicily & Sardinia   
1                            Avidagos      87  15.000000              Douro   
2                             unknown      87  14.000000             Oregon   
3                Reserve Late Harvest      87  13.000000           Michigan   
4  Vintner's Reserve Wild Child Block      87  65.000000             Oregon   

              region_1         taster_name taster_twitter_handle  \
0                 Etna       Kerin O’Keefe          @kerinokeefe   
1              unknown          Roger Voss            @vossroger   
2    Willamette Valley        Paul Gregutt           @paulgwine    
3  Lake Michigan Shore  Alexander Peartree               unknown   
4    Willamette Valley        Paul Gregutt           @paulgwine    

                                               title         variety  \
0                  Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1      Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2      Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3  St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4  Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   

                winery  
0              Nicosia  
1  Quinta dos Avidagos  
2            Rainstorm  
3           St. Julian  
4         Sweet Cheeks

**1. РАЗБОР ДАТЫ И ВРЕМЕНИ**

Признаки, обозначающие дату и время, могут содержать много полезной информации. 

*Например*, в нашем датасете в названии вина можно увидеть год производства вина. Этот признак может оказаться важным для определения рейтинга вина. Однако, пока он заключен в названии вина, модель не сможет его использовать.

<img src="images/Разбор даты и времени.png">

Можно выделить год, месяц и часы в отдельные признаки: время суток, времена года и так далее. И если вы считаете, что ваш процесс зависит от времени, посвятите этому этапу как можно больше усилий.

Подробнее про создание признаков из даты и времени вы узнаете в одном из следующих юнитов.

**2. РАЗБОР ЧИСЛОВЫХ ВЕЛИЧИН**

В наборе данных очень часто встречаются числовые величины. Это могут быть вес, рост, расстояние и так далее. Вы можете преобразовывать их, создавать новые признаки, которые могут оказаться лучше исходных.

Например, в датасете винных обзоров числовым признаком является цена за бутылку вина. Вы можете округлить цену **35.363389 $** в **35 $,** избавив данные от лишней информации.

Создадим новый признак **price_round,** означающий округлённую до целого числа цену за бутылку вина:

In [4]:
# Для удобства сразу преобразуем признак в int
df_wine_cleared["price_round"] = df_wine_cleared["price"].round().astype(int)
df_wine_cleared.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points      price           province  \
0                        Vulkà Bianco      87  35.363389  Sicily & Sardinia   
1                            Avidagos      87  15.000000              Douro   
2                             unknown      87  14.000000             Oregon   
3                Reserve Late Harvest      87  13.000000           Michigan   
4  Vintner's Reserve Wild Child Block      87  65.000000             Oregon   

              region_1         taster_name taster_twitter_handle  \
0                 Etna       Kerin O’Keefe          @kerinokeefe   
1              unknown          Roger Voss            @vossroger   
2    Willamette Valley        Paul Gregutt           @paulgwine    
3  Lake Michigan Shore  Alexander Peartree               unknown   
4    Willamette Valley        Paul Gregutt           @paulgwine    

                                               title         variety  \
0                  Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1      Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2      Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3  St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4  Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   

                winery  price_round  
0              Nicosia           35  
1  Quinta dos Avidagos           15  
2            Rainstorm           14  
3           St. Julian           13  
4         Sweet Cheeks           65

Для таких преобразований важен контекст. В контексте вина для сомелье цена в **35 $** и **35.363389 $** одинакова. Вы делаете округление признака, чтобы модель также не сосредотачивалась на сотых. Это может помочь улучшить предсказание.

Однако такое преобразование неуместно в ситуациях, когда сотые важны. *Например*, при предсказании курса лиры стоимость валюты в **6.12 руб.** и в **6.84 руб.** - абсолютно разные ситуации. 

Помимо округления для создания новых признаков могут применяться такие популярные операции, как *логарифмирование числового признака* или *извлечение его корня*. Это подразумевает создание признаков в наиболее удобной форме для обучения модели. О необходимости таких преобразований мы узнаем далее в юните преобразования данных.

**3. РАЗБОР ТЕКСТА**

Текстовые признаки могут содержать в себе большое количество информации. *Например*, если бы наша модель умела распознавать текст, то из описания вина мы бы могли понять, какими словами сомелье чаще характеризуют качественное вино. Вы можете воспользоваться методами извлечения слов из текста, чтобы получить новые признаки.

В названии вина вы можете заметить указание года произведения вина. Чтобы узнать, влияет ли год на рейтинг вина, вам необходимо выделить год в отдельный признак.

Для того, чтобы извлечь из строки число, обозначающее год, необходимо воспользоваться **регулярными выражениями.**

**Регулярные выражения** *(regexp, или regex)* - это механизм для поиска и замены текста. Это шаблоны, которые используются для поиска соответствующей части текста.

*Например*, с помощью такого регулярного выражения **[^@ \t\r\n]+@[^@ \t\r\n]+\.[^@ \t\r\n]+** можно найти любой **e-mail** в тексте.

<img src="images/Демонстрация работы регулярного выражения для поиска email в тексте.png">

Синтаксис регулярных выражений необычен и содержит различные символы, знаки препинания, математических операций и так далее.

Реализация такого механизма существует в *Pandas* в работе со строками. Для того чтобы найти все числа в каждом значении серии, воспользуемся методом **str.findall().** Метод возвращает все совпадения с заданным шаблоном в серии *Pandas*. 

Выполним код для нахождения года вина при помощи регулярного выражения, полученный признак **year** будет выглядеть следующим образом:

In [5]:
regex = '\d{4}' # регулярное выражение для нахождения чисел
df_wine_cleared["year"] = df_wine_cleared["title"].str.findall(regex).str.get(0)
df_wine_cleared.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points      price           province  \
0                        Vulkà Bianco      87  35.363389  Sicily & Sardinia   
1                            Avidagos      87  15.000000              Douro   
2                             unknown      87  14.000000             Oregon   
3                Reserve Late Harvest      87  13.000000           Michigan   
4  Vintner's Reserve Wild Child Block      87  65.000000             Oregon   

              region_1         taster_name taster_twitter_handle  \
0                 Etna       Kerin O’Keefe          @kerinokeefe   
1              unknown          Roger Voss            @vossroger   
2    Willamette Valley        Paul Gregutt           @paulgwine    
3  Lake Michigan Shore  Alexander Peartree               unknown   
4    Willamette Valley        Paul Gregutt           @paulgwine    

                                               title         variety  \
0                  Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1      Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2      Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3  St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4  Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   

                winery  price_round  year  
0              Nicosia           35  2013  
1  Quinta dos Avidagos           15  2011  
2            Rainstorm           14  2013  
3           St. Julian           13  2013  
4         Sweet Cheeks           65  2012

Разберём регулярное выражение **\d{4}:**

* **\d -** класс символов, обозначает соответствие цифрам в диапазоне цифр **[0-9];**
* **{4}** в шаблоне означает искать четыре вхождения символа, указанного ранее. В нашем случае это будут четырехзначные числа.

Таким образом, **\d{4}** означает поиск четырехзначных чисел в заданной строке.

Однако при поиске числа методом **data['title'].str.findall(regex)** результатом выполнения является список найденных цифр. Поэтому необходимо извлечь первый элемент из списка найденных методом **str.get(0),** где **0 -** первый элемент в списке найденных чисел.

Используя методы работы со строками *Рandas* и регулярные выражения, можно извлечь любую информацию для новых признаков: даты, слова и выражения.

**4. РАЗБОР КАТЕГОРИЙ**

В наборе данных винных обзоров самая популярная страна-производитель вина - *США*. Возможно, это не случайность, и факт производства в США влияет на рейтинг вина. Выделим этот факт.

Мы можем создать новый бинарный признак **is_usa** и присвоить ему **1** в случае, если вино произведено в *США*, иначе - **0.**

In [6]:
df_wine_cleared["is_usa"] = df_wine_cleared["country"].apply(lambda x: 1 if x == "US" else 0)
df_wine_cleared.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points      price           province  \
0                        Vulkà Bianco      87  35.363389  Sicily & Sardinia   
1                            Avidagos      87  15.000000              Douro   
2                             unknown      87  14.000000             Oregon   
3                Reserve Late Harvest      87  13.000000           Michigan   
4  Vintner's Reserve Wild Child Block      87  65.000000             Oregon   

              region_1         taster_name taster_twitter_handle  \
0                 Etna       Kerin O’Keefe          @kerinokeefe   
1              unknown          Roger Voss            @vossroger   
2    Willamette Valley        Paul Gregutt           @paulgwine    
3  Lake Michigan Shore  Alexander Peartree               unknown   
4    Willamette Valley        Paul Gregutt           @paulgwine    

                                               title         variety  \
0                  Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1      Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2      Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3  St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4  Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   

                winery  price_round  year  is_usa  
0              Nicosia           35  2013       0  
1  Quinta dos Avidagos           15  2011       0  
2            Rainstorm           14  2013       1  
3           St. Julian           13  2013       1  
4         Sweet Cheeks           65  2012       1

В наборе данных также есть ещё две страны, которые являются не менее популярными производителями вина.

**1.** Выберем из списка две самых популярных (помимо США) страны, производящих вино.

In [7]:
# Найдем самые популярные страны
popular_countries = df_wine_cleared["country"].value_counts()
print(popular_countries)

country
US                        54504
France                    22093
Italy                     19540
Spain                      6645
Portugal                   5691
Chile                      4472
Argentina                  3800
Austria                    3345
Australia                  2329
Germany                    2165
New Zealand                1419
South Africa               1401
Israel                      505
Greece                      466
Canada                      257
Hungary                     146
Bulgaria                    141
Romania                     120
Uruguay                     109
Turkey                       90
Slovenia                     87
Georgia                      86
England                      74
Croatia                      73
Mexico                       70
Moldova                      59
Brazil                       52
Lebanon                      35
Morocco                      28
Peru                         16
Ukraine                      14


Возможно, производство вина в этих странах также влияет на рейтинг. Давайте создадим для них бинарные признаки.

**2.** Создадим бинарные признаки **is_france, is_italy** наподобие признака **is_usa.**

In [8]:
df_wine_cleared["is_france"] = df_wine_cleared["country"].apply(lambda x: 1 if x == "France" else 0)
df_wine_cleared["is_france"].sum()

22093

In [9]:
df_wine_cleared["is_italy"] = df_wine_cleared["country"].apply(lambda x: 1 if x == "Italy" else 0)
df_wine_cleared["is_italy"].sum()

19540

Мы можем углубиться в разбор категориальных признаков - создать бинарный признак для каждого значения страны. Можно не делать это вручную: существуют специальные способы кодирования. Об особенностях этого подхода мы поговорим в юните *Кодирование данных*.

Также мы можем создавать некие бизнес-правила, гипотезы. О правилах их построения мы узнаем далее в материалах, посвящённых статистическим гипотезам. 

Попробуем вывести правило в определении качества вина: старые вина лучше молодых. В нашем датасете **40 %** вин старше **2010 года.** 

**3.** Создадим новый бинарный признак **old_wine,** где значение **1** будет обозначать, что вино старше **2010 года.**

In [10]:
# Преобразуем признак year в объект datetime для удобного сравнения дат
df_wine_cleared["year"] = pd.to_datetime(df_wine_cleared["year"], errors="coerce")

# Для сравнения используем год, заполняем значения признака old_wine, где год вина меньше 2010
df_wine_cleared["old_wine"] = df_wine_cleared["year"].apply(lambda x: 1 if x.year < 2010 else 0)
df_wine_cleared["old_wine"].sum()

39765

Итак, разбор категориальных признаков может быть использован для выделения самых популярных, уникальных категорий в отдельный признак.

Таким образом, разбор категорий, даты и времени, числовых признаков, текста применяются для создания новых, более сильных признаков для обучения.

Выполним следующие задания для закрепления изученного материала.

**4.** *Создание признаков -* это …

* процесс добавления новых признаков на основе исходного набора данных или из внешних источников данных

**5.** *Проектирование признаков* включает в себя … 

* создание признаков;
* преобразование признаков;
* отбор признаков.

**6.** Сопоставим этапы создания признаков и используемые приёмы.

**1.** извлечение бинарного признака для значения категории - это ...

* разбор категорий

**2.** использование логарифмирования, округления - это ...

* разбор числовых значений

**3.** извлечение количества дней из разницы дат - это ...

* разбор даты и времени

**4.** использование регулярных выражений - это ...

* разбор текста

**5.** работа со строковыми методами *Pandas* - это ...

* разбор текста

**7.** СоздаДИМ новый признак *locality* из признака *title*, который будет обозначать название *долины/местности производства вина*.

*Например*, в названии вина *Rainstorm 2013 Pinot Gris (Willamette Valley) locality* будет *Willamette Valley*. В названии *Tandem 2011 Ars In Vitro Tempranillo-Merlot (Navarra) - Navarra*.

Ответом на это задание является получившееся регулярное выражение и код преобразования.

In [14]:
regex = '\((.*?)\)'
df_wine_cleared["locality"] = df_wine_cleared["title"].str.findall(regex).str.get(0)
df_wine_cleared.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points      price           province  \
0                        Vulkà Bianco      87  35.363389  Sicily & Sardinia   
1                            Avidagos      87  15.000000              Douro   
2                             unknown      87  14.000000             Oregon   
3                Reserve Late Harvest      87  13.000000           Michigan   
4  Vintner's Reserve Wild Child Block      87  65.000000             Oregon   

              region_1         taster_name taster_twitter_handle  \
0                 Etna       Kerin O’Keefe          @kerinokeefe   
1              unknown          Roger Voss            @vossroger   
2    Willamette Valley        Paul Gregutt           @paulgwine    
3  Lake Michigan Shore  Alexander Peartree               unknown   
4    Willamette Valley        Paul Gregutt           @paulgwine    

                                               title         variety  \
0                  Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1      Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2      Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3  St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4  Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   

                winery  price_round       year  is_usa  is_france  is_italy  \
0              Nicosia           35 2013-01-01       0          0         1   
1  Quinta dos Avidagos           15 2011-01-01       0          0         0   
2            Rainstorm           14 2013-01-01       1          0         0   
3           St. Julian           13 2013-01-01       1          0         0   
4         Sweet Cheeks           65 2012-01-01       1          0         0   

   old_wine             locality  
0         0                 Etna  
1         0                Douro  
2         0    Willamette Valley  
3         0  Lake Michigan Shore  
4         0    Willamette Valley

В этом юните мы рассмотрели основные подходы к созданию признаков из существующего набора данных. Далее мы узнаем, как создавать новые признаки из внешних источников данных. 

#### **III. Создание признаков. Внешние источники данных**